[APM] Add MRF support on trace-agent

ichinaski opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

Jira ticket: https://datadoghq.atlassian.net/browse/APMSP-1459

Initial commit to add MRF support on trace agent:
- Read MRF configuration from Remote Config client, and propagate the changes to the core config.
- Reads multi_region_failover configuration  on Trace Agent initialization and adds MRF endpoints when applicable.
- Keep track of MRF endpoint states, and enables its senders when failover has been enabled.

Notes:

* This implementation replicates what the Logs and Metrics Agents are doing to support MRF. Most of the implementation details piggyback on the existing logic for these agents.
* While the Trace Agent handles more endpoints in its API (reverse proxies for EvP, telemetry, etc), this PR and this stage of the MRF rollout is scoped to Trace and Stats payloads only. Other reverse proxies will *not* be failed over to a different DC.
* Remote Configuration already had the [required schema](https://github.com/DataDog/dd-go/blob/44da58db0a74caba0177672593f960b8279d2980/remote-config/apps/rc-schema-validation/schemas/agent-failover.json) changes for APM. This PR honors that configuration key: `multi_region_failover.failover_apm`.
* Failover status is checked on every payload send, in case the internal state has changed. When not enabled (default state), the MRF senders will simply disregard the payload.

### Motivation

https://datadoghq.atlassian.net/browse/APMSP-1459

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

Unit tests hande internal testing.

TODO: Verify whether e2d tests exist for MRF in other agents.

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

ichinaski committed:
[APM] Add MRF support on trace-agent

Initial commit to add MRF support on trace agent:
- Reads multi_region_failover configuration and adds MRF endpoints when
  applicable
- Keep track of MRF endpoint states, and enables its senders when
  failover has been enabled.

TODO:
- Tests
- Ensure how endpoints other than traces, stats should be handled on MRF

ichinaski committed:
clean up linter errors

ichinaski committed:
Propagate RC changes for MRF on APM

ichinaski committed:
Add release notes

ichinaski committed:
Merge branch 'main' into ichinaski/apm-mrf

ichinaski committed:
add unit test for MRF senders

ajgajg1134 commented on pkg/trace/writer/sender.go:
```diff
@@ -46,14 +46,16 @@ func newSenders(cfg *config.AgentConfig, r eventRecorder, path string, climit, q
 			os.Exit(1)
 		}
 		senders[i] = newSender(&senderConfig{
```
Maybe we shouldn't consider evenly splitting the max conns on line 39 for MRF endpoints as this could limit the number of connections we use when MRF is not active (which I assume would be the default state?)

ichinaski replied:
well spotted! I will fix this, we should definitely keep max_conns for the expected senders only (MRF's default state is indeed not being active)

drichards-87 commented on releasenotes/notes/apm-add-mrf-config-48e0220faa45c703.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+features:
+  - |
+    APM: Add Multi-Region failover support. MRF is controlled through Remote
+    Config, and allows the trace-agent to switch to a failover DC when enabled.
```
```suggestion
    APM: Add support for multi-region failover. This feature is controlled via Remote Config and allows the trace-agent to switch to a failover data center when enabled.
```

drichards-87 reviewed (approved):
Left a suggestion from Docs and approved the PR.

ichinaski committed:
Update releasenotes/notes/apm-add-mrf-config-48e0220faa45c703.yaml

Co-authored-by: DeForest Richards <56796055+drichards-87@users.noreply.github.com>

ajgajg1134 commented on pkg/trace/writer/sender.go:
```diff
@@ -326,6 +334,32 @@ func (s *sender) recordEvent(t eventType, data *eventData) {
+	if s.cfg.isMRFEnabled() {
+		if !s.enabled {
+			log.Infof("Sender for domain %v has been failed over to, enabling it for MRF.", s.cfg.url)
+			s.enabled = true
```
I think we can reduce the level of nesting here by setting s.enabled and then just returning true here?

ichinaski committed:
Fix max conns initialization on writer senders

Do not take MRF endpoints into account when spreading out the maximum
number of connections.

ichinaski commented on pkg/trace/writer/sender_test.go:
```diff
@@ -230,6 +230,58 @@ func TestSender(t *testing.T) {
+		}
+
+		// Server 1 receives payloads
+		assert.Equal(20, servers[0].Total(), "total")
```
hmm I may be interpreting the meaning the other way around, but isn't the current implementation expected, actual already?
eg: `20` is the expected number we know, `servers[0].Total()` the actual value we get from the test. 🤔

ichinaski committed:
clean up nested code

heyronhay reviewed (approved):
Looks good from the DDR/HAMR/MRF side of things!

remeh commented on comp/trace/config/setup.go:
```diff
@@ -176,6 +183,22 @@ func applyDatadogConfig(c *config.AgentConfig, core corecompcfg.Component) error
+	// Add MRF endpoint, if enabled
+	if core.GetBool("multi_region_failover.enabled") {
+		prefix := apiEndpointPrefix + mrfPrefix
+		mrfURL, err := utils.GetMRFEndpoint(core, prefix, "multi_region_failover.dd_url")
```
Nevermind, I've checked the implementation and `GetMRFEndpoint` seems to take care of this 🙇

ichinaski committed:
Clean up TODO comment

ichinaski commented:
/merge

ichinaski commented:
/merge
